doc/man/julia-install.1.md
# julia-install 1 "Nov 2019" julia-install "User Manuals"

## SYNOPSIS

`julia-install` [[JULIA-VERSION | JULIA [VERSION]] [-- CONFIGURE_OPTS...]]

## DESCRIPTION

Installs Julia.

https://github.com/jlenv/julia-install#readme

## ARGUMENTS

*JULIA*
  Install Julia by name. Must be `julia`.

*VERSION*
  Optionally select the version of selected Julia.

*CONFIGURE_OPTS*
  Additional optional configure arguments.

## OPTIONS

`-r`, `--julias-dir` *DIR*
    Specifies the alternate directory where other Julia directories are
    installed, such as *~/.jlenv/versions*.
    Defaults to */opt/julias* for root and *~/.julias* for normal users.

`-i`, `--install-dir` *DIR*
    Specifies the directory where Julia will be installed.
    Defaults to */opt/julias/$julia-$version* for root and
    *~/.julias/$julia-$version* for normal users.

`--prefix` *DIR*
    Alias for `-i DIR`.

`--system`
    Alias for `-i /usr/local`.

`-s`, `--src-dir` *DIR*
    Specifies the directory for downloading and unpacking Julia source.

`-c`, `--cleanup`
    Remove the downloaded Julia archive and unpacked source-code after
    installation.

`-j[`*JOBS*`]`, `--jobs[=`*JOBS*`]`
    Specifies the number of *make* jobs to run in parallel when compiling
    Julia. If the -j option is provided without an argument, *make* will
    allow an unlimited number of simultaneous jobs.

`-p`, `--patch` *FILE*
    Specifies any additional patches to apply.

`-M`, `--mirror` *URL*
    Specifies an alternate mirror to download the Julia archive from.

`-u`, `--url` *URL*
    Alternate URL to download the Julia archive from.

`-g`, `--gpg` *GnuPG*
    Specifies to use the GPG signature for the Julia archive.

`--package-manager [apt|dnf|yum|pacman|zypper|brew|pkg|port]`
  Use an alternative package manager.

`--no-download`
    Use the previously downloaded Julia archive.

`--no-verify`
    Do not verify the downloaded Julia archive.

`--no-extract`
    Do not extract the downloaded Julia archive. Implies `--no-download`
    and `--no-verify`.

`--no-install-deps`
    Do not install build dependencies before installing Julia.

`--no-reinstall`
    Skip installation if another Julia is detected in same location.

`-L`, `--latest`
    Downloads the latest julia versions and checksums from the julia-versions
    repository (https://github.com/jlenv/julia-versions#readme).

`-V`, `--version`
    Prints the current julia-install version.

`-h`, `--help`
    Prints a synopsis of julia-install usage.

## EXAMPLES

List supported Julias and their major versions:

    $ julia-install

List the latest version:

    $ julia-install --latest

Install the current stable version of Julia:

    $ julia-install julia

Install the latest version of Julia:

    $ julia-install --latest julia

Install a latest version of Julia:

    $ julia-install julia 1.2

Install a specific version of Julia:

    $ julia-install julia 1.0.1

Install a Julia into a specific directory:

    $ julia-install --install-dir /path/to/dir julia

Install a Julia into a specific `julias` directory:

    $ julia-install --julias-dir /path/to/julias/ julia

Install a Julia into `/usr/local`:

    $ julia-install --system julia 1.0.1

Install a Julia from an alternate site with directly download:

    $ julia-install -M https://ftp.julialang.org/pub/julia julia 1.0.1

Install a Julia from a mirror:

    $ julia-install -M http://www.mirrorservice.org/sites/ftp.julialang.org/pub/julia julia 1.0.1

Install a Julia with a specific patch:

    $ julia-install -p https://raw.githubusercontent.com/JuliaLang/julia/master/deps/patches/clang-D28477.patch julia 1.0.3-p28477

Install a Julia with specific configuration:

    $ julia-install julia 1.0.1 -- --enable-shared --enable-dtrace CFLAGS="-O3"

Using julia-install for use with with [jlenv]:

    $ julia-install -i ~/.jlenv/versions/1.0.1 julia 1.0.1

Uninstall a Julia version:

    $ rm -rf ~/.julias/julia-1.0.1

## FILES

*/usr/local/src*
    Default root user source directory.

*~/src*
    Default non-root user source directory.

*/opt/julias/$julia-$version*
    Default root user installation directory.

*~/.julias/$julia-$version*
    Default non-root user installation directory.

## AUTHOR

Postmodern <postmodern.mod3@gmail.com>
Mark Van de Vyver <mark@taqtiqa.com>

## SEE ALSO

julia(1), jlenv(1), chjulia(1), chjulia-exec(1)


docs/changelog.md
### 0.1.0 / 2019-10-31

Initial Julia release.

* Supports installing arbitrary versions.
* Supports downloading the latest versions and checksums from [julia-versions].
* Verifies GPG signatures of Julia archives.
* Compatible with `chjulia` and the `jlenv` plugin `jlenv-chjl`.
* Supports default install to XDG Base Directories [specification][xdg-spec]
  * User directories
    * `XDG_CONFIG_HOME`
      Where user-specific configurations should be written (analogous to /etc).
      Default: $HOME/.config/julias/.
    * `XDG_CACHE_HOME`
       Where user-specific non-essential (cached) data should be written (analogous to /var/cache).
       Default: $HOME/.cache/julias/.
    * `XDG_DATA_HOME`
       Where user-specific data files should be written (analogous to /usr/share).
       Default: $HOME/.local/share/julias/.
  * System directories
    * `XDG_DATA_DIRS`
       List of directories seperated by : (analogous to PATH).
       Default: /usr/local/share/julias/:/usr/share/julias/.
    * `XDG_CONFIG_DIRS`
       List of directories seperated by : (analogous to PATH).
       Default: /etc/xdg/julias/.
* Supports installing into arbitrary directories.
* Supports downloading from arbitrary URLs.
* Supports downloading from mirrors.
* Supports downloading/applying patches.
* Supports specifying arbitrary `./configure` options.
* Supports downloading archives using `wget` or `curl`.
* Supports verifying downloaded archives using `md5sum`, `md5` or `openssl md5`.
* Supports installing build dependencies via the package manager:
  * [apt]
  * [dnf]
  * [yum]
  * [pacman]
  * [zypper]
  * [pkg]
  * [macports]
  * [brew]
* Has tests.

[xdg-spec]: https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html

[apt]: http://wiki.debian.org/Apt
[dnf]: https://fedoraproject.org/wiki/Features/DNF
[yum]: http://yum.baseurl.org/
[pacman]: https://wiki.archlinux.org/index.php/Pacman
[brew]: http://mxcl.github.com/homebrew/

[Julia]: http://www.julialang.org/

[julia-versions]: https://github.com/jlenv/julia-versions#readme


share/julia-install/julia-install.sh
#!/usr/bin/env bash

shopt -s extglob

julia_install_version="0.3.0"
julia_install_dir="${BASH_SOURCE[0]%/*}"
julia_install_cache_dir="${XDG_CACHE_HOME:-$HOME/.cache}/src"

julias=(julia)
patches=()
configure_opts=()
make_opts=()

system_dir="/usr/local"

if (( UID == 0 )); then
  # The src_dir are XDG locations.
	src_dir="${julia_install_cache_dir}"
	julias_dir="/opt/julias"
else
  # The src_dir are XDG locations.
	src_dir="${julia_install_cache_dir}"
	julias_dir="$HOME/.julias"
fi

. "${julia_install_dir}/util.sh"
. "${julia_install_dir}/julia-versions.sh"
. "${julia_install_dir}/functions.sh"

#
# Prints usage information for julia-install.
#
function usage()
{
	cat <<USAGE
usage: julia-install [OPTIONS] [JULIA [VERSION] [-- CONFIGURE_OPTS ...]]

Options:

	-r, --julias-dir DIR	Directory that contains other installed Julias
	-i, --install-dir DIR	Directory to install Julia into
	    --prefix DIR        Alias for -i DIR
	    --system		Alias for -i $system_dir
	-s, --src-dir DIR	Directory to download source-code into
	-c, --cleanup		Remove archive and unpacked source-code after installation
	-j, --jobs JOBS		Number of jobs to run in parallel when compiling
	-p, --patch FILE	Patch to apply to the Julia source-code
	-M, --mirror URL	Alternate mirror to download the Julia archive from
	-u, --url URL		Alternate URL to download the Julia archive from
	-g, --gpg		Verify GPG signature of the Julia archive
	-e  --ed		Verify Minisign/Signify signature of the Julia archive
	--no-download		Use the previously downloaded Julia archive
	--no-verify		Do not verify the downloaded Julia archive
	--no-extract		Do not re-extract the downloaded Julia archive
	--no-install-deps	Do not install build dependencies before installing Julia
	--no-reinstall  	Skip installation if another Julia is detected in same location
	-L, --latest		Downloads the latest julia versions and signatures
	-V, --version		Prints the version
	-h, --help		Prints this message

Examples:

	$ julia-install julia
	$ julia-install julia 1.0
	$ julia-install julia 1.0.4
	$ julia-install julia -- --with-openssl-dir=...
	$ julia-install -M https://ftp.julialang.org/pub/julia julia
	$ julia-install -M http://www.mirrorservice.org/sites/ftp.julialang.org/pub/julia julia
	$ julia-install -p https://raw.github.com/gist/4136373/falcon-gc.diff julia 1.9.3

USAGE
}

#
# Parses a "julia-version" string.
#
function parse_julia()
{
	local arg="$1"

	case "$arg" in
		*-*)
			julia="${arg%%-*}"
			julia_version="${arg#*-}"
			;;
		*)
			julia="${arg}"
			julia_version=""
			;;
	esac

	if [[ ! "${julias[@]}" == *"$julia"* ]]; then
		error "Unknown julia: $julia"
		return 1
	fi
}

#
# Parses command-line options for julia-install.
#
function parse_options()
{
	local argv=()

	while [[ $# -gt 0 ]]; do
		case $1 in
			-r|--julias-dir)
				julias_dir="$2"
				shift 2
				;;
			-i|--install-dir|--prefix)
				install_dir="$2"
				shift 2
				;;
			--system)
				install_dir="$system_dir"
				shift 1
				;;
			-s|--src-dir)
				src_dir="$2"
				shift 2
				;;
			-c|--cleanup)
				cleanup=1
				shift
				;;
			-j|--jobs|-j+([0-9])|--jobs=+([0-9]))
				make_opts+=("$1")
				shift
				;;
			-p|--patch)
				patches+=("$2")
				shift 2
				;;
			-M|--mirror)
				julia_mirror="$2"
				shift 2
				;;
			-u|--url)
				julia_url="$2"
				shift 2
				;;
			-g|--gpg)
				julia_gpg=1
				shift
				;;
			--ed)
				julia_ed=1
				shift
				;;
			--no-download)
				no_download=1
				shift
				;;
			--no-verify)
				no_verify=1
				shift
				;;
			--no-extract)
				no_download=1
				no_verify=1
				no_extract=1
				shift
				;;
			--no-install-deps)
				no_install_deps=1
				shift
				;;
			--no-reinstall)
				no_reinstall=1
				shift
				;;
			-L|--latest)
				force_update=1
				shift
				;;
			-V|--version)
				echo "julia-install: $julia_install_version"
				exit
				;;
			-h|--help)
				usage
				exit
				;;
			--)
				shift
				configure_opts=("$@")
				break
				;;
			-*)
				echo "julia-install: unrecognized option $1" >&2
				return 1
				;;
			*)
				argv+=($1)
				shift
				;;
		esac
	done

	case ${#argv[*]} in
		2)	parse_julia "${argv[0]}-${argv[1]}" || return $? ;;
		1)	parse_julia "${argv[0]}" || return $? ;;
		0)	return 0 ;;
		*)
			echo "julia-install: too many arguments: ${argv[*]}" >&2
			usage 1>&2
			return 1
			;;
	esac
}

#
# Prints Julias supported by julia-install.
#
function list_julias()
{
	local julia

	for julia in "${julias[@]}"; do
		if [[ $force_update -eq 1 ]] ||
		   are_julia_versions_missing "$julia"; then
			log "Downloading latest $julia versions ..."
			download_julia_versions "$julia"
		fi
	done

	echo "Stable julia versions:"
	for julia in "${julias[@]}"; do
		echo "  $julia:"
		stable_julia_versions "$julia" | sed -e 's/^/    /' || return $?
	done
}

#
# Initializes variables.
#
function init()
{
	local fully_qualified_version="$(lookup_julia_version "$julia" "$julia_version")"

	if [[ -n "$fully_qualified_version" ]]; then
		julia_version="$fully_qualified_version"
	else
		warn "Unknown $julia version $julia_version. Proceeding anyways ..."
	fi

	julia_cache_dir="${julia_install_cache_dir}/${julia}"
	install_dir="${install_dir:-$julias_dir/$julia-$julia_version}"

	source "$julia_install_dir/functions.sh"       || return $?
	source "$julia_install_dir/$julia/functions.sh" || return $?

	julia_gpg="${julia_gpg:-$(julia_signature_for "$julia" gpg "$julia_archive")}"
}
